### DISCUSSION TITLE
Modifications to taxonomy to support implementation of IfcOffsetCurveByDistance

### DISCUSSION THREAD

**Discussion by RickBrice** [IfcOpenShell/IfcOpenShell#4190] [Q&A]
@aothms I'm looking to re-implement IfcOffsetCurveByDistance because it doesn't work as well as I'd like. My original assumption was that all offset curves would be offset by something that maps to taxonomy::piecewise_function. Since the basis curve can be any IfcCurve, this is a bad assumption.

It looks like a lot of the curves, like IfcPolyline, IfcIndexedPolyCurve, and others map to taxonomy::loop.

To apply the piecewise offsets, and to project the piecewise offset to the head and tail of the basis curve, the length of the basis curve is needed. The offsets are linearly interpolated between instances and get projected to the head and tail of the curve if the first/last offset are defined after the start or before the end of the curve.

Any objections to adding a length() method to taxonomy::edge and taxonomy::loop to make it easy to get the total length of a curve that maps to a loop?

Thought I'd ask before I start messing up your code.

**Comment by aothms** [comment]
> Any objections to adding a length() method to taxonomy::edge and taxonomy::loop to make it easy to get the total length of a curve that maps to a loop?

This surely sounds reasonable and rest assured you have carte blanche to mess up anything you see :)

The only real problem I foresee is things like b-spline curves where we probably can't compute the length in the mapping layer (because we don't want to reimplement a nurbs library there).

And maybe that would also be an alternative to consider, to not implement IfcOffsetCurveByDistances in the mapping layer, but in the geometry library layer, where we have more geometric tooling. That would mean just creating a new taxonomy type the just hold the data verbatim. I'm fine either way though, because moving this down to opencascade just to evaluate bspline lengths (which aren't used anyway in IFC, just happen to exist in the schema) it's going to incur a massive performance hit.

The distinction between loop, edge and curve is unfortunately also really fuzzy and is for a large part due to polyline, which is a subtype of curve, but should be mapped to a loop. We have the same with surface and face, in IFC you have can have a surface_of_linear_extrusion of a polyline, which should be mapped to a shell because of the discontinuities, but is a subtype of surface in IFC.

**Comment by RickBrice** [comment]
@aothms What do you mean by "geometry library layer" versus the "mapping layer"? Perhaps I'm missing a fundamental concept.

I just discovered this morning that there is a [bSI-InfraRoom](https://github.com/bSI-InfraRoom) with a [TI-Bridge-Documentation](https://github.com/bSI-InfraRoom/TI-Bridge-Documentation) repository. From what I understand from https://bsi-infraroom.github.io/TI-Bridge-Documentation/4_3_0_0/bridge-test-instructions/HTML/link/ifcalignment.htm, IfcOffsetCurveByDistances can be the geometric representation of an IfcAlignment. That detail did not come through from the official schema documentation and concept templates.

This makes me look at IfcOffsetCurveByDistances differently. I'm also looking at IfcPolyline differently when used for an alignment.

Perhaps having a subclass of piecewise_function, maybe called loop_function, that has a pointer to a loop and treats the loop as a piecewise function will work better. The use case that comes to mind is resolving the geometry of an IfcLinearPlacement where the basis curve is an IfcPolyline or an IfcOffsetCurveByDistances derived from and IfcPolyline. To get the XYZ point for the IfcLinearPlacement, the basis curve needs to be evaluated at IfcLinearPlacement.RelativePlacement.DistanceAlong. This is what piecewise_function::evaluate(double u) does. loop_function::evaluate(u) could override this for the underlying loop object.

I guess there isn't a question here - just me thinking out loud and releasing a little frustration. Any insights you have are appreciated.

**Reply by aothms** [reply]
> IfcOffsetCurveByDistances can be the geometric representation of an IfcAlignment. That detail did not come through from the official schema documentation and concept templates.

Here is a similar conclusion:

https://github.com/buildingSMART/ifc-gherkin-rules/blob/main/features/ALS005_Alignment-shape-representation.feature#L15

> resolving the geometry of an IfcLinearPlacement where the basis curve is an IfcPolyline or an IfcOffsetCurveByDistances derived from and IfcPolyline

I don't consider this a highly likely use case. From the MVD you referenced we read for example:

> which may consist of a polyline **for capturing existing conditions** at approximate intervals.

In those cases is it really likely that you're going to position off of this?

I think we can also simply add a 0th and 1st derivative to loop and edge directly (or just call it evaluate()), rather than subclassing and overriding in a novel class. We can special case that in the linear placement evaluation and later see if we need to reshuffle the inheritance hierarchy once we have a more complete picture.

> What do you mean by "geometry library layer" versus the "mapping layer"? Perhaps I'm missing a fundamental concept.

An example is for example boolean_result, where nothing geometrically is computed in the mapping layer, the taxonomy just stores the operand and defers processing to the geometry layer.

- https://github.com/IfcOpenShell/IfcOpenShell/blob/v0.8.0/src/ifcgeom/mapping/IfcBooleanResult.cpp#L40
- https://github.com/IfcOpenShell/IfcOpenShell/blob/v0.8.0/src/ifcgeom/kernels/opencascade/boolean_result.cpp#L46
- https://github.com/IfcOpenShell/IfcOpenShell/blob/v0.8.0/src/ifcgeom/kernels/cgal/CgalKernel.cpp#L1727

**Comment by RickBrice** [comment]
> "geometry library layer" versus the "mapping layer"

I understand. "Geometry layer" is CGal, OpenCascade, etc... I was thinking it was the ifcopenshell::geometry namespace in the c++ code.

I have come up with a much less intrusive solution. The mapping for IfcPointByDistanceExpression and IfcOffsetCurveByDistances both need to be able to evaluate the basis curve at some DistanceAlong. Depending on the type of the basis curve, the mapping could be loop or piecewise_function. Only piecewise_function has the evaluate(distalong) method.

Solution: create a piecewise_function from a loop:
In profile_helper.h/cpp

~~~
taxonomy::piecewise_function::ptr ifcopenshell::geometry::piecewise_from_item(taxonomy::item::ptr item) {
    taxonomy::piecewise_function::ptr pwf;

    pwf = ifcopenshell::geometry::taxonomy::dcast<ifcopenshell::geometry::taxonomy::piecewise_function>(item);
    if (pwf) {
        return pwf; // item is a piecewise_function, just return it
    }

    auto loop = taxonomy::dcast<ifcopenshell::geometry::taxonomy::loop>(item);
    if (loop) {
	// item is a loop. create a piecewise_function from its linear edges
        pwf = taxonomy::make<taxonomy::piecewise_function>();
	  for (auto& edge : loop->children)
	  {
            const auto& s = boost::get<taxonomy::point3::ptr>(edge->start)->ccomponents();
            const auto& e = boost::get<taxonomy::point3::ptr>(edge->end)->ccomponents();
            Eigen::Vector3d v = e - s;
            auto l = v.norm(); // the norm of a vector is a measure of its length
            v.normalize(); // normalize the vector so that it is a unit direction vector
            std::function<Eigen::Matrix4d(double)> fn = [s, v](double u) 
		{
                Eigen::Vector3d o(s+u*v), axis(0, 0, 1), refDirection(v);
                auto Y = axis.cross(refDirection).normalized();
                axis = refDirection.cross(Y).normalized();
                return taxonomy::make<taxonomy::matrix4>(o, axis, refDirection)->components();
	       };
            pwf->spans.emplace_back(l, fn);
	  }

	  return pwf;
    }

	 Logger::Error(std::exception("Unexpected item type when creating piecewise_function"));
    return pwf;
}
~~~

Usage example for mapping of IfcPointByDistanceExpression

~~~
taxonomy::ptr mapping::map_impl(const IfcSchema::IfcPointByDistanceExpression* inst) {
   auto u = (*inst->DistanceAlong()->as<IfcSchema::IfcLengthMeasure>()) * length_unit_;
   auto basis_curve = inst->BasisCurve();
   auto item = map(basis_curve);
   auto pw_curve = ifcopenshell::geometry::piecewise_from_item(item);
   auto m = pw_curve->evaluate(u);
...
~~~

The big problem with `piecewise_from_item` is dealing with cases when `edge` is a curve (e.g. trimmed_curve.basis != nullptr). I'll put some more effort into this before committing.

**Reply by aothms** [reply]
Great!

> The big problem with piecewise_from_item is dealing with cases when edge is a curve (e.g. trimmed_curve.basis != nullptr). I'll put some more effort into this before committing.

I wouldn't spend too much time on this now. IFC is a huge beast to implement exactly because of these combinatorial possibilities. Just throw an exception and deal with it later once we have concrete demand.

This reminds me of another mechanism btw that we have in this taxonomy thing. Upgrades:

https://github.com/IfcOpenShell/IfcOpenShell/blob/142b15e3322a366818166832573636a486f4d120/src/ifcgeom/taxonomy.h#L1073-L1097

And these upgrades (currently only one) are actually automatically applied when trying to cast.

https://github.com/IfcOpenShell/IfcOpenShell/blob/142b15e3322a366818166832573636a486f4d120/src/ifcgeom/taxonomy.h#L1103

I imagine your loop to piecewise "upgrade" could also be moved here and maybe also that allows for a way to centrally handle caching because I can imagine that computing this anew for every IfcPointByDistanceExpression is potentially costly (but also not a real concern for me now at the point being).

**Reply by RickBrice** [reply]
I'll look at loop_to_face_upgrade to see if I can do something similar.

BTW - I found the IFC documentation that lists all the geometric representations for an IfcAlignment. My frustration with that was unfounded.

**Comment by RickBrice** [comment]
> I imagine your loop to piecewise "upgrade" could also be moved here 

It can and it's a great solution.

What does this line accomplish? Is it something applicable only to `face`?
https://github.com/IfcOpenShell/IfcOpenShell/blob/142b15e3322a366818166832573636a486f4d120/src/ifcgeom/taxonomy.h#L1081

**Reply by aothms** [reply]
Yes, a face can have multiple bounds, one outer, zero or more inner. So if you upgrade a loop to a face, it's per definition the outer bound. There's some of these things such as external and orientation (even orientation_2) that I'd still like to refine.

**Comment by RickBrice** [comment]
Thanks - I won't mess with `loop->external` since it doesn't apply to piecewise_function.

I'd like to get a better understanding of orientation_2. I think it represents the same concept as the ParentCurve placement and the IfcCurveSegment placement for IfcCompositeCurve.  I still have many questions but ended up hacking together a solution that seems to produce the right geometry.

So, on a related note, I volunteered to present what I've been doing in IfcOpenShell regarding alignments in the Implementors Forum meeting. I'll present next week (1/18/2024). One thing I'll discuss is placement and trimming of parent curves for IfcCurveSegment. Hopefully that will generate some good conversation.

**Reply by aothms** [reply]
> I'd like to get a better understanding of orientation_2. I think it represents the same concept as the ParentCurve placement and the IfcCurveSegment placement for IfcCompositeCurve

Not sure. It's specifically for this:

If you have a vanilla ifc4 composite curve, you have two layers where you toggle the orientation of the underlying curve, with slightly different behaviour. Probably the correct way would be to store one on the curve (so you have a reversed circle basically) and one on the edge.

- IfcCompositeCurveSegment.SameSense

  https://ifc43-docs.standards.buildingsmart.org/IFC/RELEASE/IFC4x3/HTML/lexical/IfcCompositeCurveSegment.htm

- IfcTrimmedCurve.SenseAgreement

  https://ifc43-docs.standards.buildingsmart.org/IFC/RELEASE/IFC4x3/HTML/lexical/IfcTrimmedCurve.htm

I also would like to store this orientation outside of the actual taxonomy item, but rather embed it in the shared_ptr that references it. Because it's actually more of a quality of how the item is used as opposed to the identity of the item itself and it makes it easier to assert that edges have exactly one neighbour in a manifold shell. But as you can read this is all very much about the non-infra geometries I wouldn't bother you too much with the details.

> So, on a related note, I volunteered to present what I've been doing in IfcOpenShell regarding alignments in the Implementors Forum meeting. I'll present next week (1/18/2024).

That's great!

**Comment by RickBrice** [comment]
> I imagine your loop to piecewise "upgrade" could also be moved here and maybe also that allows for a way to centrally handle caching because I can imagine that computing this anew for every IfcPointByDistanceExpression is potentially costly

I think I've made an assumption that caching already happens in loop_to_face_upgrade, so it would come for free to loop_to_piecewise_function_upgrade. That isn't the case, is it?

To implement a caching scheme, loop would need to be revised with a boost::optional<taxonomy::piecewise_function::ptr> to hold the upgrade object. The loop_to_piecewise_function_upgrade class would then check the loop for a previously created PWF and use it, otherwise create a cache. Am I thinking about this as you envision?


Sorry - we have two ideas in this discussion, but I wanted to say that there are a lot of similarities between IfcCompositeCurveSegment.SameSense and the sign of IfcCurveSegment.SegmentLength. The sign of SegmentLength defines sense agreement. IfcCompositeCurve.Segments is where vanilla geometry and infra geometry seem to diverge. The types of IfcSegment are different. Not sure why that is. For now, we can think about if there are enough similarities to have common code for IfcCompositeCurve and the related IfcSegment.

**Reply by aothms** [reply]
> To implement a caching scheme, loop would need to be revised with a boost::optionaltaxonomy::piecewise_function::ptr to hold the upgrade object

Yes, or we need to make `(d?)cast<>()` a member function of `class mapping` and store the previously upgraded items in a map. I think I would find that less intrusive.

**Comment by RickBrice** [comment]
@aothms I implemented by caching on the `loop` class. f12c3aae09f0bb8a981ffdab19a39c2b32a6e5ed

If you think this is too intrusive, I'll rework this by putting the `(d?)cast<>()` as a member of the `class mapping`

**Reply by aothms** [reply]
Perfect. If needed I can do that later.
